streams/src/main/java/org/apache/kafka/streams/state/internals/TimeOrderedKeySchema.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.kstream.Window;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.kstream.internals.TimeWindow;
+import org.apache.kafka.streams.state.StateSerdes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/**
+ * A {@link RocksDBSegmentedBytesStore.KeySchema} to serialize/deserialize a RocksDB store
+ * key into a schema combined of (time,seq,key). This key schema is more efficient when doing
+ * range queries between a time interval. For key range queries better use {@link WindowKeySchema}.
+ */
+public class TimeOrderedKeySchema implements RocksDBSegmentedBytesStore.KeySchema {
+    private static final Logger LOG = LoggerFactory.getLogger(TimeOrderedKeySchema.class);
+
+    private static final int TIMESTAMP_SIZE = 8;
+    private static final int SEQNUM_SIZE = 4;
+    private static final int PREFIX_SIZE = TIMESTAMP_SIZE + SEQNUM_SIZE;
+
+    @Override
+    public Bytes upperRange(final Bytes key, final long to) {
+        final byte[] maxPrefix = ByteBuffer.allocate(PREFIX_SIZE)
+            .putLong(to)
+            .putInt(Integer.MAX_VALUE)
+            .array();
+
+        return OrderedBytes.range(maxPrefix, key);
+    }
+
+    @Override
+    public Bytes lowerRange(final Bytes key, final long from) {
+        final byte[] minPrefix = ByteBuffer.allocate(PREFIX_SIZE)
+            .putLong(from)
+            .putInt(0)
+            .array();
+
+        return OrderedBytes.range(minPrefix, key);
+    }
+
+    @Override
+    public Bytes upperRangeFixedSize(final Bytes key, final long to) {
```

Comment:
We know these functions are triggered by `fetch(final Bytes key, final long timeFrom, final long timeTo)` and following the default implementation it is sub-optimal since we will range over a large scan and then drop a lot of the records. Let's add a oneliner comment on top of them referring readers to the head javadoc of this schema class that they should try avoid ever calling these functions.